### PR TITLE
=per #16659 fix deserialization of SnapshotHeader AND SWITCH TO Scala 2.11.4

### LINKED ISSUE #16659: bridge Snapshot deserialization for Scala 2.11
`scala.Option` changed its serialVersionUID between 2.10 and 2.11, but the change itself is completely harmless. We can retain the ability to read SnapshotHeader objects by working around the UID alone.

### PR DESCRIPTION
**Pull request by rkuhn**
builds on top of #16575, fixes #16659

### CODE REVIEW

**Comment by ktoso** [comment]
Awesome patch, scary but really important for users I think. Awaiting tests :)

**Comment by rkuhn** [comment]
Added test for 2.11 Java-style serialized snapshots (the comment about the `Some(...)` case was nonsensical as SnapshotSerializer never stores a manifest) and also fixed downgrading to 2.10 (i.e. the inverted patch). Now this should be it—with Jenkins’ permission.

**Comment by rkuhn** [comment]
PLS BUILD

(failure was due to missing genjavadoc plugin for Scala 2.11.5)

**Comment by bantonsson** [comment]
PLS BUILD

File: akka-persistence/src/main/scala/akka/persistence/serialization/SnapshotSerializer.scala [hunk 1]
```diff
@@ -118,6 +143,57 @@ class SnapshotSerializer(system: ExtendedActorSystem) extends Serializer {
...
+  // The offset of the serialVersionUID in the above.
+  val offset = 16
+  // This is the new serialVersionUID from Scala 2.11 onwards.
+  val replacement: Array[Byte] = Array(
+    0xfe, 0x69, 0x37, 0xfd, 0xdb, 0x0e, 0x66, 0x74).map(_.toByte)
+  val doPatch: Boolean = {
+    val baus = new ByteArrayOutputStream()
+    val out = new ObjectOutputStream(baus)
+    out.writeObject(None)
+    val superOffset = 30
```
Comment (drewhk) [inline akka-persistence/src/main/scala/akka/persistence/serialization/SnapshotSerializer.scala]: buh, can you put some comments what is happening here? Expecially with the array copying down below.

File: akka-persistence/src/main/scala/akka/persistence/serialization/SnapshotSerializer.scala [hunk 2]
```diff
@@ -87,28 +87,53 @@ class SnapshotSerializer(system: ExtendedActorSystem) extends Serializer {
   private def snapshotFromBinary(bytes: Array[Byte]): AnyRef = {
     val extension = SerializationExtension(system)
 
-    val headerLength = readInt(new ByteArrayInputStream(bytes))
+    val in = new ByteArrayInputStream(bytes)
+    val headerLength = readInt(in)
     val headerBytes = bytes.slice(4, headerLength + 4)
     val snapshotBytes = bytes.drop(headerLength + 4)
 
+    def patch(b: Array[Byte]): Array[Byte] = {
```
Comment (drewhk) [inline akka-persistence/src/main/scala/akka/persistence/serialization/SnapshotSerializer.scala]: please add some comments what is happening here.

**Comment by drewhk** [comment]
otherwise LGTM

**Comment by ktoso** [comment]
I found this not to work recently, including on `2.11.4` which it originally passed with;

```
> akka-persistence-experimental > ++2.11.4 testOnly *SnapshotSerializerPersistenceSpec*
[info] - must be able to read snapshot created with akka 2.3.6 *** FAILED ***
[info]   java.lang.ClassNotFoundException: sr-akka.persistence.serialization.SnapshotHeaderI
                                                                                            serializerIdmanifesttLscala/Option;xpsr
                                                                                                                                   scala.None$FP$?S??xr
                                                                                                                                                       scala.Option?`$?2?E?xp
[info]   at java.lang.Class.forName0(Native Method)
[info]   at java.lang.Class.forName(Class.java:249)
[info]   at akka.actor.ReflectiveDynamicAccess$$anonfun$getClassFor$1.apply(DynamicAccess.scala:67)
[info]   at akka.actor.ReflectiveDynamicAccess$$anonfun$getClassFor$1.apply(DynamicAccess.scala:66)
[info]   at scala.util.Try$.apply(Try.scala:191)
[info]   at akka.actor.ReflectiveDynamicAccess.getClassFor(DynamicAccess.scala:66)
[info]   at akka.persistence.serialization.SnapshotSerializer$$anonfun$1.apply(SnapshotSerializer.scala:113)
[info]   at akka.persistence.serialization.SnapshotSerializer$$anonfun$1.apply(SnapshotSerializer.scala:113)
[info]   at scala.Option.map(Option.scala:145)
[info]   at akka.persistence.serialization.SnapshotSerializer.snapshotFromBinary(SnapshotSerializer.scala:113)
[info]   ...
```

Did not yet uncover when exactly it started failing, will look into it after the webinar..

**Pull request merged**